sprint_8-solution-time-and-duration

7Jenko opened the pull request:
Добавить новые поля и методы согласно ТЗ 8.

7Jenko committed:
sprint_8-solution-time-and-duration
Добавить новые поля и методы согласно ТЗ 8.

7Jenko committed:
sprint_8-solution-time-and-duration
Добавить новые поля и методы согласно ТЗ 8.

7Jenko committed:
sprint_8-solution-time-and-duration
Добавить новые поля и методы согласно ТЗ 8.

slavrikk commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -5,16 +5,16 @@
     protected final Map<Integer, Subtask> subtasks = new HashMap<>();
     protected final Map<Integer, Epic> epics = new HashMap<>();
     private final HistoryManager historyManager = Managers.getDefaultHistory();
+    protected TreeSet<Task> prioritizedTasks = new TreeSet<>();
```
Лучше использовать интерфейс в качестве ссылки `Set<Task> prioritizedTasks = new TreeSet<>();`

slavrikk commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -194,26 +209,79 @@ public List<Task> getHistory() {
+                .reduce(Duration.ZERO, Duration::plus);
+    }
+
+    private void updatePrioritizedTasks() {
```
Лучше просто добавлять новую таску в коллекцию `prioritizedTasks`, не вычитывать все таски из коллекций заново

slavrikk commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -194,26 +209,79 @@ public List<Task> getHistory() {
+        List<Subtask> subtasks = getSubtasks();
+        tasks.addAll(subtasks);
+        List<Task> tasksWithStartTime = tasks.stream()
+                .filter(task -> task.getStartTime() != null)
```
Если нужно учесть null в getStartTime(), можно укзать дополнительную логику в компараторе

slavrikk commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -194,26 +209,79 @@ public List<Task> getHistory() {
+                            .plus(task1.getDuration())));
+                });
+        if (overlapTime) {
+            throw new ManagerSaveException("Задачи пересекаются по времени");
```
Лучше не кидать эксепшион, а просто не добавлять таску

slavrikk commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -194,26 +209,79 @@ public List<Task> getHistory() {
+        }
+        boolean overlapTime = prioritizedTasks.stream()
+                .anyMatch(task1 -> {
+                    LocalDateTime endTime = task.getStartTime().plus(task.getDuration());
```
В классе Task можно создать метод `getEndTime();` , который будет считаться как `startTime.plus(duration);` , чтобы не повторять код по расчету

7Jenko committed:
sprint_8-solution-time-and-duration
Использовать интерфейс в качестве ссылки Set<Task> prioritizedTasks = new TreeSet<>();
Добавлять новую задачу в коллекцию prioritizedTasks, не вычитывать все задачи из коллекций;
Убрать null в getStartTime(), так как ТЗ предполагает не учитывать задачи без даты начала в списке задач;
Убрать exception при пересечении задач по времени. Исправить тест;
Создать метод getEndTime().
